Add support for AirPods mute API

youennf opened the pull request:
#### 277f1203fdad2dce4ff1af259a0bc9b1ac6bad06
<pre>
Add support for AirPods mute API
<a href="https://rdar.apple.com/135545311">rdar://135545311</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=279358">https://bugs.webkit.org/show_bug.cgi?id=279358</a>

Reviewed by Eric Carlson.

We introduce a web preference called UseMicrophoneMuteStatusAPI to control this feature.
When it is on and we are capturing microphone, we register listeners to AirPods mute notifications.
This allows to mute/unmute microphone when user presses AirPods stems.

To implement this, we register for the listeners in the GPUProcess via CoreAudioSharedUnit.
On mute/unmute notification, we send an IPC message to GPUProcessProxy.
GPUProcessProxy selects the web page that should process the notification (currently, it is the last microphone capturing web page).
The web page will then toggle the microphone muted state according the notification.

This will in turn instruct the GPUProcess via the WebProcess to mute capture which is needed on macOS.
On iOS, microphone is muted by the system directly, on macOS this is done via kAUVoiceIOProperty_MuteOutput.
This patch needs <a href="https://bugs.webkit.org/show_bug.cgi?id=279515">https://bugs.webkit.org/show_bug.cgi?id=279515</a> so that, when muting, we keep the audio unit to be able to recveive the unmute notifications.

Manually tested on macOS and iOS.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/cocoa/AVFAudioSoftLink.h: Added.
* Source/WebCore/PAL/pal/cocoa/AVFAudioSoftLink.mm: Added.
* Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.cpp:
(WebCore::CoreAudioSharedUnit::~CoreAudioSharedUnit):
(WebCore::CoreAudioSharedUnit::handleMuteStatusChangedNotification):
* Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.h:
* Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.mm:
(-[WebCoreAudioInputMuteChangeListener start]):
(-[WebCoreAudioInputMuteChangeListener stop]):
(-[WebCoreAudioInputMuteChangeListener handleMuteStatusChangedNotification:]):
(WebCore::getSharedAVAudioApplication):
(WebCore::setNoopInputMuteStateChangeHandler):
(WebCore::registerAudioInputMuteChangeListener):
(WebCore::unregisterAudioInputMuteChangeListener):
(WebCore::CoreAudioSharedUnit::setMuteStatusChangedCallback):
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::enableMicrophoneMuteStatusAPI):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::enableMicrophoneMuteStatusAPI):
(WebKit::GPUProcessProxy::updateCaptureAccess):
(WebKit::GPUProcessProxy::setMicrophoneMuteStatusChanged):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.messages.in:
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::syncWithWebCorePrefs const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setMuted):
(WebKit::WebPageProxy::willStartCapture):
(WebKit::WebPageProxy::setMicrophoneMuteStatusChanged):
(WebKit::WebPageProxy::gpuProcessExited):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: <a href="https://commits.webkit.org/284123@main">https://commits.webkit.org/284123@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ae03133bf01eeddb6156a4d32efc56632d6bac40

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68520 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47912 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21179 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/72589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19665 "Failed to checkout and rebase branch from PR 33469") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70637 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55708 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19481 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/72589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/19665 "Failed to checkout and rebase branch from PR 33469") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71587 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43799 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59173 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/72589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40466 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16596 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/18022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/61638 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/62421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16943 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74283 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/67768 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12491 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/16195 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/62135 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12530 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59252 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/62160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15188 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10094 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/3719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/89547 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43713 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15854 "Found 9 new JSC stress test failures: wasm.yaml/wasm/function-tests/many-arguments-to-function.js.wasm-eager-jettison, wasm.yaml/wasm/fuzz/memory.js.wasm-eager, wasm.yaml/wasm/fuzz/memory.js.wasm-no-cjit, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.default-wasm, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-bbq, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-eager-jettison, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-no-cjit, wasm.yaml/wasm/stress/tail-call.js.wasm-eager, wasm.yaml/wasm/stress/tail-call.js.wasm-slow-memory (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44787 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45981 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44529 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.mm:
```diff
@@ -38,7 +38,101 @@
+    setNoopInputMuteStateChangeHandler(audioApplication, true);
+#endif
+
+    auto *center = [NSNotificationCenter defaultCenter];
```
Nit: this local variable is unnecessary

youennf replied:
ok

eric-carlson commented on Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.mm:
```diff
@@ -38,7 +38,101 @@
+    setNoopInputMuteStateChangeHandler(audioApplication, false);
+#endif
+
+    auto *center = [NSNotificationCenter defaultCenter];
```
Ditto

youennf replied:
ok

eric-carlson commented on Source/WebKit/GPUProcess/GPUProcess.cpp:
```diff
@@ -426,6 +426,16 @@ void GPUProcess::setUseSCContentSharingPicker(bool use)
+#if PLATFORM(COCOA)
+    CoreAudioSharedUnit::unit().setMuteStatusChangedCallback([weakProcess = WeakPtr { *this }] (bool isMuting) {
+        if (RefPtr process = weakProcess.get())
+            process->protectedParentProcessConnection()->send(Messages::GPUProcessProxy::SetMicrophoneMuteStatusChanged(isMuting), 0);
```
Here and elsewhere: **Set**MicrophoneMuteStatus**Changed** is a confusing name. I would drop either `Set` or `Changed` - `setMicrophoneMuteStatus` or `microphoneMuteStatusChanged`

youennf replied:
Right, `microphoneMuteStatusChanged` is better

eric-carlson commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8514,6 +8514,11 @@ void WebPageProxy::setMuted(WebCore::MediaProducerMutedStateFlags state, FromApp
         state.add(WebCore::MediaProducer::MediaStreamCaptureIsMuted);
 
+#if ENABLE(MEDIA_STREAM) && ENABLE(GPU_PROCESS)
+    if (preferences().captureAudioInGPUProcessEnabled() && internals().mutedState.contains(WebCore::MediaProducerMutedState::AudioCaptureIsMuted) && !state.contains(WebCore::MediaProducerMutedState::AudioCaptureIsMuted))
```
Should we make it impossible to enable `UseMicrophoneMuteStatusAPI` when `captureAudioInGPUProcessEnabled` is false?

youennf replied:
Actually, it would be better to allow it.
I'll work on enabling this in a follow-up, this patch is already big enough.

youennf replied:
I filed https://bugs.webkit.org/show_bug.cgi?id=280235

youennf committed:
Add support for AirPods mute API
rdar://135545311
https://bugs.webkit.org/show_bug.cgi?id=279358

Reviewed by Eric Carlson.

We introduce a web preference called UseMicrophoneMuteStatusAPI to control this feature.
When it is on and we are capturing microphone, we register listeners to AirPods mute notifications.
This allows to mute/unmute microphone when user presses AirPods stems.

To implement this, we register for the listeners in the GPUProcess via CoreAudioSharedUnit.
On mute/unmute notification, we send an IPC message to GPUProcessProxy.
GPUProcessProxy selects the web page that should process the notification (currently, it is the last microphone capturing web page).
The web page will then toggle the microphone muted state according the notification.

This will in turn instruct the GPUProcess via the WebProcess to mute capture which is needed on macOS.
On iOS, microphone is muted by the system directly, on macOS this is done via kAUVoiceIOProperty_MuteOutput.
This patch needs https://bugs.webkit.org/show_bug.cgi?id=279515 so that, when muting, we keep the audio unit to be able to recveive the unmute notifications.

Manually tested on macOS and iOS.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/cocoa/AVFAudioSoftLink.h: Added.
* Source/WebCore/PAL/pal/cocoa/AVFAudioSoftLink.mm: Added.
* Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.cpp:
(WebCore::CoreAudioSharedUnit::~CoreAudioSharedUnit):
(WebCore::CoreAudioSharedUnit::handleMuteStatusChangedNotification):
* Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.h:
* Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.mm:
(-[WebCoreAudioInputMuteChangeListener start]):
(-[WebCoreAudioInputMuteChangeListener stop]):
(-[WebCoreAudioInputMuteChangeListener handleMuteStatusChangedNotification:]):
(WebCore::getSharedAVAudioApplication):
(WebCore::setNoopInputMuteStateChangeHandler):
(WebCore::registerAudioInputMuteChangeListener):
(WebCore::unregisterAudioInputMuteChangeListener):
(WebCore::CoreAudioSharedUnit::setMuteStatusChangedCallback):
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::enableMicrophoneMuteStatusAPI):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::enableMicrophoneMuteStatusAPI):
(WebKit::GPUProcessProxy::updateCaptureAccess):
(WebKit::GPUProcessProxy::setMicrophoneMuteStatusChanged):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.messages.in:
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::syncWithWebCorePrefs const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setMuted):
(WebKit::WebPageProxy::willStartCapture):
(WebKit::WebPageProxy::setMicrophoneMuteStatusChanged):
(WebKit::WebPageProxy::gpuProcessExited):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: https://commits.webkit.org/284123@main
